tsne/bh_sne_src/tsne.cpp

```diff
@@ -706,16 +714,60 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
             "%f)!\nLearning embedding...\n",
             (float)(end - start) / CLOCKS_PER_SEC,
             (double)row_P[N] / ((double)N * (double)N));
-  start = clock();
 
-  for (int iter = 0; iter < max_iter; iter++) {
+  TSNEState* state = new TSNEState;
+  state->P = P;
+  state->row_P = row_P;
+  state->col_P = col_P;
+  state->val_P = val_P;
+  state->dY = dY;
+  state->uY = uY;
+  state->gains = gains;
+  TSNE* tsne = new TSNE;
+  tsne->N = N;
+  tsne->Y = Y;
+  tsne->no_dims = NDIMS;
+  tsne->theta = theta;
+  tsne->max_iter = max_iter;
+  tsne->stop_lying_iter = stop_lying_iter;
+  tsne->mom_switch_iter = mom_switch_iter;
+  tsne->state = state;
+  tsne->iter = 0;
+  tsne->total_time = 0;
+  tsne->residual_time = 0;
+
+  return tsne;
+}
+
+// Optimize t-SNE
+template <int NDIMS>
+bool run_n(int n, struct TSNE* tsne) {
```

Comment:
This really feels like it should be a method on `tsne` rather than a function.  That would let you skip some of the reference-declaring and also would allow the compiler to skip the null-checks it's inserting everywhere.

Reply:
If I do that, it's not C-api anymore. I could make it all C++-ified, but then I'll also have to craft an independent C-api... which I could do.

Reply:
It's already very, very much not a C api because it's templated.

Reply:
OK, updated with a C++ API and an `extern "C"` wrapper for cython, bindgen consumption.